docs/socketconfig.md
Socket.IO Client Configuration
==============================

As of 2015-10-25, the legacy `/sioconfig` JavaScript for retrieving connection
information is being deprecated in favor of a new API.  The purpose of this
change is to allow partitioning channels across multiple servers in order to
better handle increasing traffic.

To get the socket.io configuration for the server hosting a particular channel,
make a `GET` request to `/socketconfig/<channel name>.json`.  The response will
be a JSON object containing a list of acceptable servers to connect to, or an
error message.

Examples:

```
GET /socketconfig/test.json
200 OK

{
    "servers": [
        {
            "url": "https://localhost:8443",
            "secure": true
        },
        {
            "url": "http://localhost:1337",
            "secure": false
        },
        {
            "url": "https://local6:8443",
            "secure": true,
            "ipv6": true
        },
        {
            "url": "http://local6:1337",
            "secure": false,
            "ipv6": true
        }
    ]
}

GET /socketconfig/$invalid$.json
404 Not Found

{
    "error": "Channel \"$invalid$\" does not exist."
}
```

Each entry in the `servers` array has `"secure":true` if the connection is
secured with TLS, otherwise it it is false.  An entry with `"ipv6":true`
indicates that the server is listening on the IPv6 protocol.

You can pick any URL to connect socket.io to in order to join the specified
channel.  I recommend picking one with `"secure":true`, only choosing an
insecure connection if implementing a TLS connection is infeasible.


docs/user-settings.md
# User Preferences #

From any CyTube channel, you can click the Options link at the top of the page to open a dialog where you can change your personal preferences.  This page explains each of the available options.

## General ##

General interface preferences.

Setting | Description
--------|------------
Theme | Choose from different colorschemes for the website.
Layout | Choose from different layouts for elements on the page.  Fluid layouts will expand to fill the entire window, while compact layouts will remain the same size.  "Synchtube" layout is the same as the default layout, but mirrored.
Ignore Channel CSS | Don't load custom stylesheets for each channel.  Requires a refresh to take effect.
Ignore Channel JavaScript | Don't load custom scripts for each channel.  The Script Access tab allows you to manage your preferences on a per-channel basis, but if this setting is checked, scripts will be globally disallowed and you will not be prompted to accept them when joining a channel.

## Playback ##

Preferences for video playback and the playlist.

Setting | Description
--------|------------
Synchronize video playback | By default, CyTube attempts to synchronize the video so that everyone is watching at the same time.  Some users with poor internet connections may wish to disable this in order to prevent excessive buffering due to constantly seeking forward.
Synch threshold | The number of seconds your video is allowed to be ahead/behind before it is forcibly seeked to the correct position.  Should be set to at least 2 seconds to avoid buffering problems and choppy playback.
Set wmode=transparent | There's probably no reason to touch this unless you know what you're doing.  Having a non-transparent wmode can cause modals to display behind the video player, but also can cause performance issues in some situations.
Remove the video player | Automatically remove the video player on page load.  Equivalent to manually clicking Layout->Remove Video every time you load a channel.
Hide playlist buttons by default | Hides the control buttons from each video in the playlist, so that only the title is displayed.  The control buttons can be shown by right clicking the video item in the playlist.
Old style playlist buttons | Legacy feature introduced in CyTube 2.0 for those who preferred the old 1.0-style video control buttons.
Quality Preference | Sets the preferred quality for player types that support quality selection (currently, this is YouTube, Vimeo, Dailymotion, Google Drive, and Google+).  If your preferred quality is not available, the next lowest quality will be used.

## Chat ##

Preferences for the integrated chatroom.

Setting | Description
--------|------------
Show timestamps in chat | When enabled, a timestamp is prepended to each chat message.  For example, `[09:45:10] message here`.
Sort userlist by rank | Controls whether the username list is sorted alphabetically and by rank, or just alphabetically.
Sort AFKers to bottom | When enabled, usernames of AFK users will be sorted to the bottom of the username list.
Blink page title on new messages | Controls the conditions under which the tab title blinks between the channel title and `*Chat*` when a new message arrives.
Notification sound on new messages | Controls the conditions under which a notification sound is played when a new message arrives.
Add a send button to chat | Adds a clickable button to send chat messages.  Only really useful for virtual keyboards that lack a dedicated Enter key.
Disable chat emotes | Disables the automatic conversion of channel-defined emote codes to inline images.

## Script Access ##

Manage your preferences for allowing or denying custom scripts for channels you've visited.  A channel will only appear here if you checked "Remember my preference" when allowing or denying a channel script.  You can toggle the preference between "Allow" and "Deny", or click "Clear Preference" to remove the saved preference, so that you will be asked every time you join the channel.

## Moderator ##

Settings that only apply to channel moderators.

Setting | Description
--------|------------
Show name color | Colors your username in chat (the same color as in the username list).  This setting is also controlled by the small button labeled "M" in the upper right corner of chat.
Show join messages | Display a message every time a user logs in to the chat.
Show shadowmuted messages | Show chat messages from shadowmuted users.  These messages will appear ~~struck through~~, and only moderators with this setting enabled will see them.


NEWS.md
2015-10-25
==========

In order to support future clustering support, the legacy `/sioconfig`
endpoint is being deprecated.  Instead, you should make a request to
`/socketconfig/<channel name>.json`.  See [the
documentation](docs/socketconfig.md) for more information.

2015-10-04
==========

  * The channel data storage system has been refactored a bit.  For
    compatibility, the default remains to store JSON objects for each channel in
    the `chandump` folder, however there is now also the option of storing
    channel data in the database.  You can take advantage of this by setting
    `channel-storage: type: 'database'` in your `config.yaml`.
    - In order to migrate existing channel data from the `chandump` files to the
      database, run `node lib/channel-storage/migrate.js`.
  * The database storage method uses foreign keys to associate the channel data
    with the corresponding row in the `channels` table.  This requires that the
    tables be stored using the InnoDB engine rather than MyISAM.  If your CyTube
    tables defaulted to MyISAM, you can fix them by running

    ```sql
    ALTER TABLE `channels` ENGINE = InnoDB;
    ```

2015-09-21
==========

  * CyTube is now transpiled with [babel] to allow the use of ES6/ES2015
    features.  All source files have been moved from `lib` to `src`.
  * Running `npm install` or `npm run postinstall` will prompt you to
    build from `src` to `lib`.
  * Running `npm run build-server` will run the build script without any
    prompts.
  * After updating with `git pull`, you should run `npm install` or `npm run
    build-server` in order to rebuild after the changes.

[babel]: https://babeljs.io/

2015-07-25
==========

  * CyTube now supports subtitles for Google Drive videos.  In order to take
    advantage of this, you must upgrade mediaquery by running `npm install
    cytube/mediaquery`.  Subtitles are cached in the google-drive-subtitles
    folder.

2015-07-07
==========

  * CyTube and CyTube/mediaquery have both been updated to use
    calzoneman/status-message-polyfill to polyfill res.statusMessage on older
    versions of node (e.g., v0.10).  After pulling, run `npm install` to update
    this dependency.  This fixes an issue where HTTP status messages from
    mediaquery were reported as `undefined`, and removes the need for manually
    looking up status messages in `lib/ffmpeg.js`.

2015-07-06
==========

  * As part of the video player rewrite, Google Drive and Google+ metadata
    lookups are now offloaded to CyTube/mediaquery.  After pulling the new
    changes, run `npm install` or `npm update` to update the mediaquery
    dependency.

  * `www/js/player.js` is now built from the CoffeeScript source files in the
    `player/` directory.  Instead of modifying it directly, modify the relevant
    player implementations in `player/` and run `npm run build-player` (or `node
    build-player.js`) to generate `www/js/player.js`.

  * Also as part of the video player rewrite, the schema for custom embeds
    changed so any custom embeds stored in the `channel_libraries` table need to
    be updated.  The automatic upgrade script will convert any custom embeds
    that are parseable (i.e., not truncated by the width of the `id` field using
    the old format) and will delete the rest (you may see a lot of WARNING:
    unable to convert xxx messages-- this is normal).  Custom embeds in channel
    playlists in the chandumps will be converted when the channel is loaded.


src/config.js
var fs = require("fs");
var path = require("path");
var Logger = require("./logger");
var nodemailer = require("nodemailer");
var net = require("net");
var YAML = require("yamljs");

var defaults = {
    mysql: {
        server: "localhost",
        port: 3306,
        database: "cytube3",
        user: "cytube3",
        password: "",
    },
    listen: [
        {
            ip: "0.0.0.0",
            port: 8080,
            http: true,
        },
        {
            ip: "0.0.0.0",
            port: 1337,
            io: true
        }
    ],
    http: {
        domain: "http://localhost",
        "default-port": 8080,
        "root-domain": "localhost",
        "alt-domains": ["127.0.0.1"],
        minify: false,
        "max-age": "7d",
        gzip: true,
        "gzip-threshold": 1024,
        "cookie-secret": "change-me"
    },
    https: {
        enabled: false,
        domain: "https://localhost",
        "default-port": 8443,
        keyfile: "localhost.key",
        passphrase: "",
        certfile: "localhost.cert",
        cafile: "",
        ciphers: "HIGH:!DSS:!aNULL@STRENGTH",
        redirect: true
    },
    io: {
        domain: "http://localhost",
        "default-port": 1337,
        "ip-connection-limit": 10
    },
    mail: {
        enabled: false,
        /* the key "config" is omitted because the format depends on the
           service the owner is configuring for nodemailer */
        "from-address": "some.user@gmail.com",
        "from-name": "CyTube Services"
    },
    "youtube-v3-key": "",
    "channel-save-interval": 5,
    "max-channels-per-user": 5,
    "max-accounts-per-ip": 5,
    "guest-login-delay": 60,
    stats: {
        interval: 3600000,
        "max-age": 86400000
    },
    aliases: {
        "purge-interval": 3600000,
        "max-age": 2592000000
    },
    "vimeo-workaround": false,
    "vimeo-oauth": {
        enabled: false,
        "consumer-key": "",
        secret: ""
    },
    "html-template": {
        title: "CyTube Beta", description: "Free, open source synchtube"
    },
    "reserved-names": {
        usernames: ["^(.*?[-_])?admin(istrator)?([-_].*)?$", "^(.*?[-_])?owner([-_].*)?$"],
        channels: ["^(.*?[-_])?admin(istrator)?([-_].*)?$", "^(.*?[-_])?owner([-_].*)?$"],
        pagetitles: []
    },
    "contacts": [
        {
            name: "calzoneman",
            title: "Developer",
            email: "cyzon@cytu.be"
        }
    ],
    "aggressive-gc": false,
    playlist: {
        "max-items": 4000,
        "update-interval": 5
    },
    "channel-blacklist": [],
    ffmpeg: {
        enabled: false,
        "ffprobe-exec": "ffprobe"
    },
    "link-domain-blacklist": [],
    setuid: {
        enabled: false,
        "group": "users",
        "user": "nobody",
        "timeout": 15
    },
    "channel-storage": {
        type: "file"
    }
};

/**
 * Merges a config object with the defaults, warning about missing keys
 */
function merge(obj, def, path) {
    for (var key in def) {
        if (key in obj) {
            if (typeof obj[key] === "object") {
                merge(obj[key], def[key], path + "." + key);
            }
        } else {
            Logger.syslog.log("[WARNING] Missing config key " + (path + "." + key) +
                        "; using default: " + JSON.stringify(def[key]));
            obj[key] = def[key];
        }
    }
}

var cfg = defaults;

/**
 * Initializes the configuration from the given YAML file
 */
exports.load = function (file) {
    try {
        cfg = YAML.load(path.join(__dirname, "..", file));
    } catch (e) {
        if (e.code === "ENOENT") {
            Logger.syslog.log(file + " does not exist, assuming default configuration");
            cfg = defaults;
            return;
        } else {
            Logger.errlog.log("Error loading config file " + file + ": ");
            Logger.errlog.log(e);
            if (e.stack) {
                Logger.errlog.log(e.stack);
            }
            cfg = defaults;
            return;
        }
    }

    if (cfg == null) {
        Logger.syslog.log(file + " is an Invalid configuration file, " +
                          "assuming default configuration");
        cfg = defaults;
        return;
    }

    var mailconfig = {};
    if (cfg.mail && cfg.mail.config) {
        mailconfig = cfg.mail.config;
        delete cfg.mail.config;
    }
    merge(cfg, defaults, "config");
    cfg.mail.config = mailconfig;

    preprocessConfig(cfg);
    Logger.syslog.log("Loaded configuration from " + file);
};

function preprocessConfig(cfg) {
    /* Detect 3.0.0-style config and warng the user about it */
    if ("host" in cfg.http || "port" in cfg.http || "port" in cfg.https) {
        Logger.syslog.log("[WARN] The method of specifying which IP/port to bind has "+
                          "changed.  The config loader will try to handle this "+
                          "automatically, but you should read config.template.yaml "+
                          "and change your config.yaml to the new format.");
        cfg.listen = [
            {
                ip: cfg.http.host || "0.0.0.0",
                port: cfg.http.port,
                http: true
            },
            {
                ip: cfg.http.host || "0.0.0.0",
                port: cfg.io.port,
                io: true
            }
        ];

        if (cfg.https.enabled) {
            cfg.listen.push(
                {
                    ip: cfg.http.host || "0.0.0.0",
                    port: cfg.https.port,
                    https: true,
                    io: true
                }
            );
        }

        cfg.http["default-port"] = cfg.http.port;
        cfg.https["default-port"] = cfg.https.port;
        cfg.io["default-port"] = cfg.io.port;
    }
    // Root domain should start with a . for cookies
    var root = cfg.http["root-domain"];
    root = root.replace(/^\.*/, "");
    cfg.http["root-domain"] = root;
    if (root.indexOf(".") !== -1 && !net.isIP(root)) {
        root = "." + root;
    }
    cfg.http["root-domain-dotted"] = root;

    // Setup nodemailer
    cfg.mail.nodemailer = nodemailer.createTransport(
        cfg.mail.config
    );

    // Debug
    if (process.env.DEBUG === "1" || process.env.DEBUG === "true") {
        cfg.debug = true;
    } else {
        cfg.debug = false;
    }

    // Strip trailing slashes from domains
    cfg.http.domain = cfg.http.domain.replace(/\/*$/, "");
    cfg.https.domain = cfg.https.domain.replace(/\/*$/, "");

    // HTTP/HTTPS domains with port numbers
    if (!cfg.http["full-address"]) {
        var httpfa = cfg.http.domain;
        if (cfg.http["default-port"] !== 80) {
            httpfa += ":" + cfg.http["default-port"];
        }
        cfg.http["full-address"] = httpfa;
    }

    if (!cfg.https["full-address"]) {
        var httpsfa = cfg.https.domain;
        if (cfg.https["default-port"] !== 443) {
            httpsfa += ":" + cfg.https["default-port"];
        }
        cfg.https["full-address"] = httpsfa;
    }


    // Socket.IO URLs
    cfg.io["ipv4-nossl"] = "";
    cfg.io["ipv4-ssl"] = "";
    cfg.io["ipv6-nossl"] = "";
    cfg.io["ipv6-ssl"] = "";
    for (var i = 0; i < cfg.listen.length; i++) {
        var srv = cfg.listen[i];
        if (!srv.ip) {
            srv.ip = "0.0.0.0";
        }
        if (!srv.io) {
            continue;
        }

        if (srv.ip === "") {
            if (srv.port === cfg.io["default-port"]) {
                cfg.io["ipv4-nossl"] = cfg.io["domain"] + ":" + cfg.io["default-port"];
            } else if (srv.port === cfg.https["default-port"]) {
                cfg.io["ipv4-ssl"] = cfg.https["domain"] + ":" + cfg.https["default-port"];
            }
            continue;
        }

        if (net.isIPv4(srv.ip) || srv.ip === "::") {
            if (srv.https && !cfg.io["ipv4-ssl"]) {
                if (srv.url) {
                    cfg.io["ipv4-ssl"] = srv.url;
                } else {
                    cfg.io["ipv4-ssl"] = cfg.https["domain"] + ":" + srv.port;
                }
            } else if (!cfg.io["ipv4-nossl"]) {
                if (srv.url) {
                    cfg.io["ipv4-nossl"] = srv.url;
                } else {
                    cfg.io["ipv4-nossl"] = cfg.io["domain"] + ":" + srv.port;
                }
            }
        }
        if (net.isIPv6(srv.ip) || srv.ip === "::") {
            if (srv.https && !cfg.io["ipv6-ssl"]) {
                if (!srv.url) {
                    Logger.errlog.log("Config Error: no URL defined for IPv6 " +
                                      "Socket.IO listener!  Ignoring this listener " +
                                      "because the Socket.IO client cannot connect to " +
                                      "a raw IPv6 address.");
                    Logger.errlog.log("(Listener was: " + JSON.stringify(srv) + ")");
                } else {
                    cfg.io["ipv6-ssl"] = srv.url;
                }
            } else if (!cfg.io["ipv6-nossl"]) {
                if (!srv.url) {
                    Logger.errlog.log("Config Error: no URL defined for IPv6 " +
                                      "Socket.IO listener!  Ignoring this listener " +
                                      "because the Socket.IO client cannot connect to " +
                                      "a raw IPv6 address.");
                    Logger.errlog.log("(Listener was: " + JSON.stringify(srv) + ")");
                } else {
                    cfg.io["ipv6-nossl"] = srv.url;
                }
            }
        }
    }

    cfg.io["ipv4-default"] = cfg.io["ipv4-ssl"] || cfg.io["ipv4-nossl"];
    cfg.io["ipv6-default"] = cfg.io["ipv6-ssl"] || cfg.io["ipv6-nossl"];

    // sioconfig
    // TODO this whole thing is messy, need to redo how the socket address is sent
    var sioconfigjson = {
        "ipv4-nossl": cfg.io["ipv4-nossl"],
        "ipv4-ssl": cfg.io["ipv4-ssl"],
        "ipv6-nossl": cfg.io["ipv6-nossl"],
        "ipv6-ssl": cfg.io["ipv6-ssl"]
    };

    var sioconfig = JSON.stringify(sioconfigjson);
    sioconfig = "var IO_URLS=" + sioconfig + ";";
    cfg.sioconfigjson = sioconfigjson;
    cfg.sioconfig = sioconfig;

    // Generate RegExps for reserved names
    var reserved = cfg["reserved-names"];
    for (var key in reserved) {
        if (reserved[key] && reserved[key].length > 0) {
            reserved[key] = new RegExp(reserved[key].join("|"), "i");
        } else {
            reserved[key] = false;
        }
    }

    /* Convert channel blacklist to a hashtable */
    var tbl = {};
    cfg["channel-blacklist"].forEach(function (c) {
        tbl[c.toLowerCase()] = true;
    });
    cfg["channel-blacklist"] = tbl;

    if (cfg["link-domain-blacklist"].length > 0) {
        cfg["link-domain-blacklist-regex"] = new RegExp(
                cfg["link-domain-blacklist"].join("|").replace(/\./g, "\\."), "gi");
    } else {
        // Match nothing
        cfg["link-domain-blacklist-regex"] = new RegExp("$^", "gi");
    }

    if (cfg["youtube-v3-key"]) {
        require("cytube-mediaquery/lib/provider/youtube").setApiKey(
                cfg["youtube-v3-key"]);
    } else {
        Logger.errlog.log("Warning: No YouTube v3 API key set.  YouTube links will " +
            "not work.  See youtube-v3-key in config.template.yaml and " +
            "https://developers.google.com/youtube/registering_an_application for " +
            "information on registering an API key.");
    }

    return cfg;
}

/**
 * Retrieves a configuration value with the given key
 *
 * Accepts a dot-separated key for nested values, e.g. "http.port"
 * Throws an error if a nonexistant key is requested
 */
exports.get = function (key) {
    var obj = cfg;
    var keylist = key.split(".");
    var current = keylist.shift();
    var path = current;
    while (keylist.length > 0) {
        if (!(current in obj)) {
            throw new Error("Nonexistant config key '" + path + "." + current + "'");
        }
        obj = obj[current];
        current = keylist.shift();
        path += "." + current;
    }

    return obj[current];
};
